exchanges/request/request.go

```diff
@@ -172,7 +173,15 @@ func (r *Requester) doRequest(ctx context.Context, endpoint EndpointLimit, newRe
 			}
 			log.Debugf(log.RequestSys, "%s request type: %s", r.name, p.Method)
 			if p.Body != nil {
-				log.Debugf(log.RequestSys, "%s request body: %v", r.name, p.Body)
+				var data []byte
+				// Below reads the body and then resets it for the actual request.
+				data, err = io.ReadAll(p.Body)
+				if err != nil {
+					log.Errorf(log.RequestSys, "%s failed to read request body: %v", r.name, err)
+				} else {
+					p.Body = bytes.NewBuffer(data)
+				}
+				log.Debugf(log.RequestSys, "%s request body: %v", r.name, string(data))
```

Comment:
Would you consider this? Si that you don't need to reset the body again?

```diff
diff --git a/exchanges/request/request.go b/exchanges/request/request.go
index a2cb626b5..bf8973644 100644
--- a/exchanges/request/request.go
+++ b/exchanges/request/request.go
@@ -1,7 +1,6 @@
 package request
 
 import (
-       "bytes"
        "context"
        "encoding/json"
        "errors"
@@ -166,22 +165,22 @@ func (r *Requester) doRequest(ctx context.Context, endpoint EndpointLimit, newRe
 
                verbose := isVerbose(ctx, p.Verbose)
 
                        log.Debugf(log.RequestSys, "%s attempt %d request path: %s", r.name, attempt, p.Path)
                        for k, d := range req.Header {
                                log.Debugf(log.RequestSys, "%s request header [%s]: %s", r.name, k, d)
                        }
                        log.Debugf(log.RequestSys, "%s request type: %s", r.name, p.Method)
-                       if p.Body != nil {
-                               var data []byte
-                               // Below reads the body and then resets it for the actual request.
-                               data, err = io.ReadAll(p.Body)
-                               if err != nil {
-                                       log.Errorf(log.RequestSys, "%s failed to read request body: %v", r.name, err)
-                               } else {
-                                       p.Body = bytes.NewBuffer(data)
+                       if req.GetBody != nil {
+                               bodyCopy, bodyErr := req.GetBody()
+                               if bodyErr != nil {
+                                       return bodyErr
+                               }
+                               payload, bodyErr := io.ReadAll(bodyCopy)
+                               if bodyErr != nil {
+                                       return bodyErr
                                }
-                               log.Debugf(log.RequestSys, "%s request body: %v", r.name, string(data))
+                               log.Debugf(log.RequestSys, "%s request body: %s", r.name, payload)
                        }
                }
```

Output:
```
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets attempt 1 request path: https://api.btcmarkets.net/v3/orders
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request header [Bm-Auth-Signature]: [WIF420WIF420WIF420WIF420WIF420WIF420WIF420WIF420WIF420WIF420]
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request header [Accept]: [application/json]
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request header [Accept-Charset]: [UTF-8]
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request header [Content-Type]: [application/json]
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request header [Bm-Auth-Apikey]: [WIF420WIF420WIF420WIF420WIF420WIF420WIF420WIF420WIF420WIF420]
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request header [Bm-Auth-Timestamp]: [1710970509087]
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request type: POST
[DEBUG]  | REQUESTER | 21/03/2024 08:35:09 | BTC Markets request body: {"amount":"1","marketId":"BTC-AUD","postOnly":true,"price":"100","side":"Bid","type":"Limit"}
```